examples/tree_source/tree_source/app.py

```diff
@@ -1,119 +1,165 @@
+    def name(self):
+        return self._icon, os.path.basename(self.path)
 
-    def add(self, entry):
-        decade = entry['year'] // 10
+    # Property that returns modified date as str
+    @property
+    def date_modified(self):
+        return datetime.fromtimestamp(self._mtime).strftime("%d %b %Y at %H:%M")
+
+    # on-demand loading of children
+    @property
+    def children(self):
+        if not self._did_start_loading:
+            self._did_start_loading = True
+            self.load_children()
+        return self._children
+
+    def load_children(self):
         try:
-            decade_root = {
-                root.decade: root
-                for root in self._decades
-            }[decade]
-        except KeyError:
-            decade_root = Decade(decade)
-            self._decades.append(decade_root)
-            self._decades.sort(key=lambda v: v.decade)
-        movie = Movie(**entry)
-        decade_root._data.append(movie)
-        self._notify('insert', parent=decade_root, index=len(decade_root._data) - 1, item=movie)
+            names = os.scandir(self.path)
+            sub_paths = [os.path.join(self.path, p) for p in names if os.path.isdir(os.path.join(self.path, p))]
+            self._children = [Node(p, self) for p in sub_paths]
+        except OSError:
+            self._children = [LoadingFailedNode(self)]
+
+        for i, child in enumerate(self._children):
+            self._notify('insert', parent=self, index=i, item=child)
+
+    def _notify(self, notification, **kwargs):
+        # pass notifications to parent
+        self.parent._notify(notification, **kwargs)
+
+    def __str__(self):
+        return os.path.basename(self.path)
+
+    def sort(self, accessor, key=None, reverse=False):
+
+        if accessor == "date_modified":  # use our own sort function
+            def sort_func(child):
+                return child._mtime
+        elif accessor == "name":
+            def sort_func(child):
+                return child.name[1].lower()
+        else:  # use the function provided by the user / default
+            def sort_func(child):
+                # sort according to value of accessor, using the provided sort key
+                try:
+                    attr = getattr(child, accessor)
+                    if isinstance(attr, tuple):
+                        icon, value = attr
+                    else:
+                        value = attr
+                    return key(value) if key else value
+                except AttributeError:
+                    return ''
+
+        # sort all children in hirarchy
+        self._children.sort(key=sort_func, reverse=reverse)
+
+        for c in self._children:
+            try:
+                c.sort(accessor, key, reverse)
+            except AttributeError:
+                pass
+
+
+class FileSystemSource(Source, Node):
+
+    def __init__(self, path='/'):
+        Source.__init__(self)
+        Node.__init__(self, path, parent=self)
```

Comment:
This looks like it's trying to do an end-run around MRO object initialization, which could have some interesting side effects. You *should* be able to get away with a single call to `super()`, provided the classes are defined in the right order, and you treat Node as a mixin (which means `FileSystemSource` will have an inheritance order of  `Node, Source`, and the constructor of `Node` consumes it's arguments and call super() in it's own constructor, even though it doesn't have a parent).

Reply:
Ah, yes, that is indeed better. Only, it requires changing ``Node._notify`` to ``Node.notify``, otherwise``super()._notify()`` will call ``Node._notify``instead of ``Source._notify``.